staging/src/k8s.io/client-go/tools/cache/reflector_data_consistency_detector.go

```diff
@@ -34,14 +34,65 @@ import (
 )
 
 var dataConsistencyDetectionForWatchListEnabled = false
+var dataConsistencyDetectionForListAgainstCacheEnabled = false
 
 func init() {
 	dataConsistencyDetectionForWatchListEnabled, _ = strconv.ParseBool(os.Getenv("KUBE_WATCHLIST_INCONSISTENCY_DETECTOR"))
+	dataConsistencyDetectionForListAgainstCacheEnabled, _ = strconv.ParseBool(os.Getenv("KUBE_LIST_AGAINST_CACHE_INCONSISTENCY_DETECTOR"))
 }
 
-type retrieveCollectedItemsFunc func() []interface{}
+type retrieveCollectedItemsFunc[U any] func() []U
 
-type listItemsFunc func(ctx context.Context, options metav1.ListOptions) (runtime.Object, error)
+type listItemsFunc[T runtime.Object] func(ctx context.Context, options metav1.ListOptions) (T, error)
+
+// CheckListAgainstCacheDataConsistencyIfRequested performs a data consistency check only when
+// the KUBE_LIST_AGAINST_CACHE_INCONSISTENCY_DETECTOR environment variable was set during a binary startup
+// for requests served from the watch-cache.
+//
+// The consistency check is meant to be enforced only in the CI, not in production.
+// The check ensures that data retrieved by a list api call against the watch-cache
+// is exactly the same as data received by the list api call against etcd.
+//
+// Note that this function will panic when data inconsistency is detected.
+// This is intentional because we want to catch it in the CI.
+func CheckListAgainstCacheDataConsistencyIfRequested[T runtime.Object](ctx context.Context, identity string, listItemsFn listItemsFunc[T], optionsUsedToReceiveList metav1.ListOptions, receivedList runtime.Object) {
+	if !dataConsistencyDetectionForListAgainstCacheEnabled || wasListRequestServedFromStorage(optionsUsedToReceiveList) {
+		return
+	}
+	receivedListMeta, err := meta.ListAccessor(receivedList)
+	if err != nil {
+		panic(err)
+	}
+	rawListItems, err := meta.ExtractListWithAlloc(receivedList)
+	if err != nil {
+		panic(err) // this should never happen
+	}
+	lastSyncedResourceVersion := receivedListMeta.GetResourceVersion()
+	checkDataConsistency(ctx, identity, lastSyncedResourceVersion, listItemsFn, func() []runtime.Object { return rawListItems })
+}
+
+// wasListRequestServedFromStorage based on the passed ListOptions determines
+// if the request was served from the underlying storage (instead of the watch-cache).
+//
+// Note that this function is unable to detect the following cases:
+//
+// if ResourceVersion = "" and ConsistendListFromCache is disabled or RequestWatchProgress isn't supported,
+// then the request will be served from the storage.
+func wasListRequestServedFromStorage(opts metav1.ListOptions) bool {
```

Comment:
copied from the cacher.go - it looks like it does what we want expect the case mentioned in the comment and after clarifying rows 1 and 3 from the KEP.